{% extends 'base.html' %}
{% load staticfiles %}

{% comment %}

   Copyright 2015, Institute for Systems Biology

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

{% endcomment %}

{% block header %}
{% endblock %}

{% block link_page_name %}help{% endblock %}
{% block page_name %}help{% endblock %}

{% block content %}
    <div class="col-lg-12">
        <h1 class="c1 c19 title">System Documentation (v1.0)</h1>

        <div class="toc">
            <p class="c3 c8"><span class="c1"><a class="c2" href="#h.798t60msx5rt">1. User Dashboard</a></span></p>
            <p class="c3 c8"><span class="c1"><a class="c2" href="#h.v24w7rc8kom1">2. Cohorts</a></span></p>
            <p class="c3 c8"><span class="c1"><a class="c2" href="#h.9w1beo3v3c46">3. Visualizations</a></span></p>
            <p class="c3 c8"><span class="c1"><a class="c2" href="#h.7im6bigprni4">4. SeqPeek</a></span></p>
            <p class="c3 c8"><span class="c1"><a class="c2" href="#h.6rnx5eq5qtaj">5. IGV Browser</a></span></p>
            <p class="c3 c8"><span class="c1"><a class="c2" href="#h.a1np5snph00b">6. Sharing</a></span></p>
            <p class="c3 c8"><span class="c1"><a class="c2" href="#h.6173glqdalig">7. General Permissions</a></span></p>
        </div>
        <div class="real-content">
            <h1 class="c0"><a name="h.798t60msx5rt"></a><span>1. User Dashboard</span></h1><h2 class="c0"><a name="h.mmxx06i7pgo9"></a><span>1.1 Features</span></h2><h3 class="c0"><a name="h.aezcmrdlw2eh"></a><span>1.1.1 Create a cohort</span></h3><p class="c3"><span>To create a cohort from the user dashboard, click on the &ldquo;+ Create&rdquo; button and select &ldquo;New Cohort&rdquo;. This will take you to the cohort creation page.</span></p><h3 class="c0"><a name="h.f6vfo3ci6bdx"></a><span>1.1.2 Create a general visualization</span></h3><p class="c3"><span>To create a visualization from the user dashboard, click on the &ldquo;+ Create&rdquo; button and select &ldquo;Visualization&rdquo;. This will take you to a new visualization with default settings selected.</span></p><h3 class="c0"><a name="h.zdbp0opvzrnv"></a><span>1.1.3 Create a SeqPeek visualization</span></h3><p class="c3"><span>To create a visualization from the user dashboard, click on the &ldquo;+ Create&rdquo; button and select &ldquo;SeqPeek&rdquo;. This will take you to a new SeqPeek visualization with no default settings selected.</span></p><h3 class="c0"><a name="h.hmjrf8sksjct"></a><span>1.1.4 Batch operations on cohorts, visualizations, and SeqPeek</span></h3><h4 class="c0"><a name="h.msgk14fqidwb"></a><span>Delete</span></h4><p class="c3"><span>To delete a set of cohorts, first check the boxes next to the cohorts you wish to delete. The &ldquo;Trash&rdquo; button should become selectable after selecting at least one cohort. Click on the &ldquo;Trash&rdquo; button to delete the selected cohorts. This is the same for Visualizations</span></p><h4 class="c0"><a name="h.2omz8f9wnm6y"></a><span>Share</span></h4><p class="c3"><span>To share a set of cohorts, first select the boxes next to the cohorts you wish to share. The &ldquo;Share&rdquo; button should become selectable after selecting at least one cohort. Click on the &ldquo;Share&rdquo; button to share the selected cohorts. A dialogue box will appear and prompt you to select the users you wish to share with. You will be able to remove cohorts you no longer want to share. You will be able to select from a list of users that are already registered in the system and you may select one or more users you wish to share with. Click the &ldquo;Share Cohort&rdquo; button when you are done. This is the same for visualizations. </span></p><p class="c3 c4"><span></span></p><p class="c3"><span>Users who you share a cohort with will only be able to view and comment on the cohort and not make changes. They will need to clone the cohort before making changes to it.</span></p><h3 class="c0"><a name="h.xfvt4lqm3pbi"></a><span>1.1.5 Set operations on cohorts</span></h3><p class="c3"><span>To activate the set operations button, you must have at least one cohort selected. Upon clicking the &ldquo;Set Operations&rdquo; button, a dialogue box will appear. Here you may do the following things:</span></p><ul class="c9 lst-kix_9ir3dhm6ymvb-0 start"><li class="c3 c5"><span>Enter in a name for the new cohort you&rsquo;re about to create.</span></li><li class="c3 c5"><span>Select a set operation.</span></li><li class="c3 c5"><span>Edit cohorts to be used in the operation.</span></li></ul><p class="c3 c4"><span></span></p><p class="c3"><span>The intersect and union operations can take any number of cohorts and in any order.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>The complement operation requires that there be a base cohort, from which the other cohorts will be subtracted from.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>Click &ldquo;Okay&rdquo; to complete the operation and create the new cohort.</span></p><h3 class="c0"><a name="h.awva01oc4efu"></a><span>1.1.6 Cohort listing</span></h3><p class="c3"><span>When the &ldquo;Cohorts&rdquo; tab is selected, the system is displaying all the cohorts that you own and all the cohorts that have been shared with you.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>Clicking on the name of a cohort in the list will take you to the cohort details and editing page. </span></p><h3 class="c0"><a name="h.qe2bqt8isjf2"></a><span>1.1.7 Visualization listing</span></h3><p class="c3"><span>When the &ldquo;Visualizations&rdquo; tab is selected, the system is displaying all the visualizations that you own and all the visualizations that have been shared with you. </span></p><p class="c3 c4"><span></span></p><h3 class="c0"><a name="h.6sdymoci0bwb"></a><span>1.1.8 SeqPeek visualization listing</span></h3><p class="c3"><span>When the &ldquo;SeqPeek&rdquo; tab is selected, the system is displaying all the SeqPeek visualizations that you own and all the SeqPeek visualization that have been shared with you.</span></p><h3 class="c0"><a name="h.bfzt22f1a4o5"></a><span>1.1.9 Search for cohorts and visualizations</span></h3><p class="c3"><span>To search for cohorts and visualizations by name, you can use the search bar at the top of the page. This will produce a results page of two lists; one for cohorts and one for visualizations. This search only does a string matching with the names of cohorts and visualizations.</span></p><h3 class="c0"><a name="h.dc2yknch26ym"></a><span>1.1.10 Sorting of cohorts, visualizations, and SeqPeek</span></h3><p class="c3"><span>You can sort the listing of cohorts and visualizations using the column headers. By clicking on a column, it will indicate whether it is sorting that column in ascending order or descending order.</span></p><p class="c3 c4"><span></span></p>
            <h1 class="c0"><a name="h.v24w7rc8kom1"></a><span>2. Cohorts</span></h1><p class="c3 c4"><span></span></p><p class="c3"><span>Cohorts is a way of creating groupings of the samples and participant data that you&rsquo;re interested in looking at. You can create cohorts that span across multiple projects, only contain certain types of data, or focus in on a particular participant group. </span></p><p class="c3 c4"><span></span></p><h2 class="c0"><a name="h.iv3vbxfja06d"></a><span>2.1 Creating and saving a cohort</span></h2><h3 class="c0"><a name="h.nnxlivcros2b"></a><span>2.1.1 Use the create button</span></h3><p class="c3"><span>To create a cohort from the user dashboard, click on the &ldquo;+ Create&rdquo; button and select &ldquo;New Cohort&rdquo;. This will take you to the cohort creation page.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>Details of new cohort page</span></p><ul class="c9 lst-kix_o4goedadlnom-0 start"><li class="c3 c5"><span>Using the provided list of filters on the left hand side, you can select what attributes and features you are interested. By clicking on a feature, the field should expand and provide you with options on which to filter on. For example, when you click on &ldquo;Vital Status&rdquo;, it expands and provides a list of &ldquo;Alive&rdquo;, &ldquo;Dead&rdquo;, &ldquo;None&rdquo; as options to choose from. When you select on, the filter should appear in the Selected Filters panel and visualizations on the page should be updated to reflect the new cohort that has been filtered by Vital Status. The numbers beside the selectable filter values reflect the number of samples that have that attribute based on all other filters that have been selected. </span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-0 start"><li class="c3 c6"><span>Donor Filter List</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-1 start"><li class="c3 c7"><span>Project</span></li><li class="c3 c7"><span>Study</span></li><li class="c3 c7"><span>Vital Study</span></li><li class="c3 c7"><span>Gender</span></li><li class="c3 c7"><span>Age At Diagnosis</span></li><li class="c3 c7"><span>Sample Type Code</span></li><li class="c3 c7"><span>Tumor Tissue Site</span></li><li class="c3 c7"><span>Histological Type</span></li><li class="c3 c7"><span>Prior Diagnosis</span></li><li class="c3 c7"><span>Pathologic State</span></li><li class="c3 c7"><span>Tumor Status</span></li><li class="c3 c7"><span>New Tumor Event After Initial Treatment</span></li><li class="c3 c7"><span>Histological Grade</span></li><li class="c3 c7"><span>Residual Tumor</span></li><li class="c3 c7"><span>Tobacco Smoking History</span></li><li class="c3 c7"><span>ICD-10</span></li><li class="c3 c7"><span>ICD-O-3 Histology</span></li><li class="c3 c7"><span>ICD-O-3 Site</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-0"><li class="c3 c6"><span>Data Type Filter List</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-1 start"><li class="c3 c7"><span>DNA Sequencing</span></li><li class="c3 c7"><span>RNA Sequencing</span></li><li class="c3 c7"><span>miRNA Sequencing</span></li><li class="c3 c7"><span>Protein</span></li><li class="c3 c7"><span>SNP CN</span></li><li class="c3 c7"><span>DNA Methylation</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-0"><li class="c3 c6"><span>Selected Filters Panel</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-1 start"><li class="c3 c7"><span>This is where selected filters are shown so there is an easy way to see what filters have been selected. </span></li><li class="c3 c7"><span>Clicking on &ldquo;Clear All&rdquo; will remove all selected filters.</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-0"><li class="c3 c6"><span>Clinical Features Panel</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-1 start"><li class="c3 c7"><span>This panel shows a list of treemaps that give a high level break of the samples for a handful of features:</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-2 start"><li class="c3 c15"><span>Disease Code</span></li><li class="c3 c15"><span>Vital Status</span></li><li class="c3 c15"><span>Sample Type</span></li><li class="c3 c15"><span>Tumor Tissue Site</span></li><li class="c3 c15"><span>Gender</span></li><li class="c3 c15"><span>Age at Initial Pathologic Diagnosis</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-1"><li class="c3 c7"><span>By using the &ldquo;Show More&rdquo; button, you can see two more tree maps available. </span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-0"><li class="c3 c6"><span>Data Availability Panel</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-1 start"><li class="c3 c7"><span>This panel shows a parallel sets graph of available data for the selected samples in the cohort. The large headers over the vertical bars are data types. Each data type is broken up into their different platforms and &ldquo;NA&rdquo; for samples that do not have that data type. The bars that flow horizontally indicate the number of samples that have that data. By hovering on a horizontal segment between the first two bars, you will see the number of data that have both those data type platforms. You can also reorder the vertical categories by dragging the headers left and right and reorder the platforms by dragging the platform names up and down.</span></li></ul><h3 class="c0"><a name="h.955516e8rxzz"></a><span>2.1.2 Set operations</span></h3><p class="c3"><span>You can create cohorts using set operations on the user dashboard page.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>To activate the set operations button, you must have at least one cohort selected. Upon clicking the &ldquo;Set Operations&rdquo; button, a dialogue box will appear. Here you may do the following things:</span></p><ul class="c9 lst-kix_6ygryrl14dhc-0 start"><li class="c3 c5"><span>Enter in a name for the new cohort you&rsquo;re about to create.</span></li><li class="c3 c5"><span>Select a set operation.</span></li><li class="c3 c5"><span>Edit cohorts to be used in the operation.</span></li></ul><p class="c3 c4"><span></span></p><p class="c3"><span>The intersect and union operations can take any number of cohorts and in any order.</span></p><ul class="c9 lst-kix_wkco3nxx8tli-0 start"><li class="c3 c5"><span>The complement operation requires that there be a base cohort, from which the other cohorts will be subtracted from.</span></li><li class="c3 c5"><span>Click &ldquo;Okay&rdquo; to complete the operation and create the new cohort.</span></li></ul><h2 class="c0"><a name="h.x0acb2l4xvui"></a><span>2.2 Editing a cohort</span></h2><h3 class="c0"><a name="h.lyb8l93ttam9"></a><span>2.2.1 Details of cohort edit page</span></h3><h4 class="c0"><a name="h.e0wrj6kbznbr"></a><span>2.2.1.1 Menu</span></h4><ul class="c9 lst-kix_57gkw0qw5ur3-0 start"><li class="c3 c5"><span>Add New Filters</span></li></ul><ul class="c9 lst-kix_57gkw0qw5ur3-1 start"><li class="c3 c6"><span>Selecting this menu item make the filters panel appear. And filters selected will be additive to any filters that have already been selected. To return to the previous view, you much either save any selected filters, or choose to cancel adding any new filters.</span></li></ul><ul class="c9 lst-kix_57gkw0qw5ur3-0"><li class="c3 c5"><span>Comments</span></li></ul><ul class="c9 lst-kix_57gkw0qw5ur3-1 start"><li class="c3 c6"><span>Selecting &ldquo;Comments&rdquo; will cause the Comments panel to appear. Here anyone who can see this cohort can comment on it. Comments are shared with anyone who can view this cohort and ordered by newest on the bottom.</span></li></ul><ul class="c9 lst-kix_57gkw0qw5ur3-0"><li class="c3 c5"><span>Make a Copy&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-0"><li class="c3 c6"><span>Making a copy will create a copy of this cohort with the same list of samples and patients and make you the owner of the copy. </span></li></ul><ul class="c9 lst-kix_m84veatyg579-0 start"><li class="c3 c5"><span>Share with Others</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-0"><li class="c3 c6"><span>This behaves similarly to on the user dashboard page. A dialogue box appears and the user is prompted to select users that are registered in the system to share the cohort with.</span></li></ul><h4 class="c0"><a name="h.9fbbopgug9ym"></a><span>2.2.1.2 Selected Filters Panel</span></h4><p class="c3"><span>This panel displays any filters that have been used on the cohort or any of its ancestors. These cannot be modified and any additional filters applied to this cohort will be appended to the list. </span></p><h4 class="c0"><a name="h.obe1in394zzw"></a><span>2.2.1.3 Details Panel</span></h4><p class="c3"><span>This panel displays the number of samples and participant in this cohort. These vary because some participants may have provided multiple samples. </span></p><p class="c3 c4"><span></span></p><p class="c3"><span>This panel also displays &ldquo;Your Permissions&rdquo; which can be either owner or reader.</span></p><h4 class="c0"><a name="h.y3oim3364u1u"></a><span>2.2.1.4 Clinical Features Panel</span></h4><p class="c3"><span>This panel shows a list of treemaps that give a high level break of the samples for a handful of features:</span></p><ul class="c9 lst-kix_yg1d6udur2l3-0 start"><li class="c3 c5"><span>Disease Code</span></li><li class="c3 c5"><span>Vital Status</span></li><li class="c3 c5"><span>Sample Type</span></li><li class="c3 c5"><span>Tumor Tissue Site</span></li><li class="c3 c5"><span>Gender</span></li><li class="c3 c5"><span>Age at Initial Pathologic Diagnosis</span></li></ul><p class="c3 c4"><span></span></p><p class="c3"><span>By using the &ldquo;Show More&rdquo; button, you can see two more tree maps available. </span></p><h4 class="c0"><a name="h.u1ij6r4r4tm6"></a><span>2.2.1.5 Data Availability Panel</span></h4><p class="c3"><span>This panel shows a parallel sets graph of available data for the selected samples in the cohort. The large headers over the vertical bars are data types. Each data type is broken up into their different platforms and &ldquo;NA&rdquo; for samples that do not have that data type. The bars that flow horizontally indicate the number of samples that have that data. By hovering on a horizontal segment between the first two bars, you will see the number of data that have both those data type platforms. You can also reorder the vertical categories by dragging the headers left and right and reorder the platforms by dragging the platform names up and down.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>&ldquo;View File List&rdquo; takes you to a new page where you can view the file list associated to the cohort you are looking at. </span></p><ul class="c9 lst-kix_2favm7oqy01c-0 start"><li class="c3 c5"><span>The file list page provides a paginated list of files available with all samples in the cohort. Here, &ldquo;available&rdquo; refers to files that have been uploaded to the ISB-CGC Google Cloud Project and that are open access data. You can use the &ldquo;Previous&rdquo; and &ldquo;Next&rdquo; to show more values in the list.</span></li><li class="c3 c5"><span>You may filter on these files if you are only interested in a specific data type and platform. Selecting a filter will update the list associated. The numbers next to the platform refers to the number of files available for that platform.</span></li><li class="c3 c5"><span>There is only one menu item available and that is the &ldquo;Download File List as CSV&rdquo;. Selecting this item will begin a download process of all the files available for the cohort, taking into account the selected Platform filters. The file contains the following information for each file:</span></li></ul><ul class="c9 lst-kix_z9i5crp2jx0n-0"><li class="c3 c6"><span>Sample Barcode</span></li><li class="c3 c6"><span>Platform</span></li><li class="c3 c6"><span>Pipeline</span></li><li class="c3 c6"><span>Data Level</span></li><li class="c3 c6"><span>File Path to the Cloud Storage Location</span></li></ul><h3 class="c0"><a name="h.ideszeiba253"></a><span>2.2.2 Commenting</span></h3><p class="c3"><span>Any user who owns or has had a cohort shared with them can comment on it. To open comments, use the menu button at the top right and select &ldquo;Comments&rdquo;. A sidebar will appear on the right side and any previously created comments will be shown.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>On the bottom of the comments sidebar, you can create a new comment and save it. It should appear at the bottom of the list of comments.</span></p><h2 class="c0"><a name="h.s8xeax5pmuph"></a><span>2.3 Deleting a cohort</span></h2><h3 class="c0"><a name="h.tdxki8e8in4f"></a><span>2.3.1 From dashboard</span></h3><p class="c3"><span>Select the cohorts that you wish to delete using the checkboxes next to the cohorts. When one or more are selected, the delete button will be active and you can then proceed to deleting them.</span></p><h3 class="c0"><a name="h.lo5efj6d6q8m"></a><span>2.3.2 From within a cohort</span></h3><p class="c3"><span>If you are viewing a cohort you created, then you can delete the cohort from the top right menu option. </span></p><h2 class="c0"><a name="h.e974ovp8erum"></a><span>2.4 From a visualization</span></h2><p class="c3"><span>To create a cohort from a visualization, you must be in plot selection mode. If you are in plot selection mode, the crosshairs icon in the top right corner of the plot panel should be blue. If it is not, click on it and it should turn blue. </span></p><p class="c3 c4"><span></span></p><p class="c3"><span>Once in plot selection mode, you can click and drag your cursor of the plot area to select the desired samples. For a cubbyhole plot, you will have to select each cubby that you are interested in. </span></p><p class="c3 c4"><span></span></p><p class="c3"><span>When your selection has been made, a small window should appear that contains a button labelled &ldquo;Save as Cohort&rdquo;. Click on this when you are ready to create a new cohort.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>Put in a name for you newly selected cohort and click the &ldquo;Save&rdquo; button.</span></p><h2 class="c0"><a name="h.rr54ls6ez4d3"></a><span>2.5 Copying a cohort</span></h2><p class="c3"><span>Copying a cohort can only be done from the cohort details page of the cohort you are want to copy.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>When you are looking at the cohort you wish to copy, select the &ldquo;Make A Copy&rdquo; item from the top right menu.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>This will take you to your copy of the cohort.</span></p><p class="c3 c4"><span></span></p>
            <h1 class="c0"><a name="h.9w1beo3v3c46"></a><span>3. Visualizations</span></h1><p class="c3 c4"><span></span></p><p class="c3"><span>Visualizations are how you can visualize the available data for the cohorts you have created. A visualization is a collection of one or more plots and each plot is configurable to show relevant data that can be shared with others.</span></p><p class="c3 c4"><span></span></p><h2 class="c0"><a name="h.xbl9tj6g7kvu"></a><span>3.1 Creating a visualization</span></h2><p class="c3"><span>You can create a new visualization from the user dashboard. Select the &ldquo;Visualization&rdquo; option from the &ldquo;+ Create&rdquo; menu. This will prompt you to name the visualization and provide at least one cohort to start with. It will automatically choose the last one you created. Click &ldquo;Create Visualization&rdquo;.</span></p><h2 class="c0"><a name="h.pd45hv1g4z4n"></a><span>3.2 Saving a visualization</span></h2><p class="c3"><span>Use the &ldquo;Save Visualizations&rdquo; option in the top right menu. It will save the visualization and all plots in their current configuration. It will not save any selections that have been made within plots.</span></p><h2 class="c0"><a name="h.yjwv3771g9lk"></a><span>3.3 Deleting a visualization</span></h2><h3 class="c0"><a name="h.jg7mi9zahtmp"></a><span>3.3.1 From User Dashboard</span></h3><p class="c3"><span>Select the visualizations that you wish to delete using the checkboxes next to the visualization. When one or more are selected, the delete button will be active and you can then proceed to deleting them.</span></p><h3 class="c0"><a name="h.obsd1jnu0jdr"></a><span>3.3.2 From Visualization</span></h3><p class="c3"><span>If you are viewing a visualization you created, then you can delete the cohort from the top right menu option. </span></p><h2 class="c0"><a name="h.jljnjjw6zr3z"></a><span>3.4 Copying a visualization</span></h2><p class="c3"><span>Copying a visualization can only be done from the cohort details page of the cohort you are want to copy.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>When you are looking at the visualization you wish to copy, select the &ldquo;Make A Copy&rdquo; item from the top right menu.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>This will take you to your copy of the visualization.</span></p><h2 class="c0"><a name="h.5x44wwjf1m1l"></a><span>3.5 Adding a plot to a visualization</span></h2><p class="c3"><span>To add a plot to a visualization, select the &ldquo;Add a Plot&rdquo; item in the top right menu.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>This will append a new plot section to your visualization with the default plot of an age histogram for the All TCGA Data Cohort.</span></p><h2 class="c0"><a name="h.nhu2da2ym2p1"></a><span>3.6 Deleting a plot from a visualization</span></h2><p class="c3"><span>To delete a plot from a visualization, select the &ldquo;Trash&rdquo; icon in the top right corner of the plot panel you wish to remove. </span></p><h2 class="c0"><a name="h.y5g9w04aw8wp"></a><span>3.7 Commenting on a plot</span></h2><p class="c3"><span>To open the comments section for a particular plot, select the &ldquo;Comment&rdquo; icon in the top right corner of the plot panel. This will cause the comment sidebar to appear from the right side of the window. </span></p><p class="c3 c4"><span></span></p><p class="c3"><span>All previous comments will be displayed with the most recent on the bottom. </span></p><p class="c3 c4"><span></span></p><p class="c3"><span>To add a new comment, type in your comment in the text box at the bottom of the panel and click the &ldquo;Comment&rdquo; button. You should see your comment appear at the bottom of the list of comments.</span></p><h2 class="c0"><a name="h.jsmrxwyralpk"></a><span>3.8 Editing a plot</span></h2><p class="c3"><span>To change the settings of a plot, select the &ldquo;Edit&rdquo; icon in the top right corner of the plot panel. This will cause the Plot Setting panel to open.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>Selecting a new Feature to plot</span></p><ul class="c9 lst-kix_a0a35a6nvs4r-0 start"><li class="c3 c5"><span>When you click on the edit icon next to the feature you would like to change (i.e. &ldquo;X Axis Feature&rdquo;, &ldquo;Y Axis Feature&rdquo;), you will be taken to the feature selection panel.</span></li><li class="c3 c5"><span>Here you must first specify the datatype of the feature you would like to plot. Each datatype requires a different set of parameters to narrow down the feature that can be used in the plot. </span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-1 start"><li class="c3 c6"><span>Clinical</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-2 start"><li class="c3 c7"><span>Single autocomplete textbox. This input searches through the names of the all the clinical features available.</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-1"><li class="c3 c6"><span>Gene Expression</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-2 start"><li class="c3 c7"><span>Gene Filter: filters down the plot-able features by a specific gene. This is an autocomplete search field for a gene name.</span></li><li class="c3 c7"><span>Platform Filter: filters down the plot-able features by platforms.</span></li><li class="c3 c7"><span>Center Filter: filters down the plot-able features by processing center.</span></li><li class="c3 c7"><span>Select Feature: provides the filtered down list of plot-able features to select from based on selected filters.</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-1"><li class="c3 c6"><span>miRNA</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-2 start"><li class="c3 c7"><span>miRNA Name Filter: filters down the plot-able features by a specific miRNA. This is an autocomplete search field for a miRNA name.</span></li><li class="c3 c7"><span>Platform Filter: filters down the plot-able features by platforms.</span></li><li class="c3 c7"><span>Value Filter: filters down the plot-able features by value.</span></li><li class="c3 c7"><span>Select Feature: provides the filtered down list of plot-able features to select from based on selected filters.</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-1"><li class="c3 c6"><span>Methylation</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-2 start"><li class="c3 c7"><span>Gene Filter: filters down the plot-able features by a specific gene. This is an autocomplete search field for a gene name.</span></li><li class="c3 c7"><span>CpG Probe Filter: filters down the plot-able features by a specific CpG Probe. This is an autocomplete search field for a particular probe.</span></li><li class="c3 c7"><span>Platform Filter: filters down the plot-able features by platforms.</span></li><li class="c3 c7"><span>Gene Region Filter: filters down the plot-able features by specific gene regions.</span></li><li class="c3 c7"><span>CpG Island region Filter: filters down the plot-able features by CpG Island region.</span></li><li class="c3 c7"><span>Select Feature: provides the filtered down list of plot-able features to select from based on selected filters.</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-1"><li class="c3 c6"><span>Copy Number</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-2 start"><li class="c3 c7"><span>Gene Filter: filters down the plot-able features by a specific gene. This is an autocomplete search field for a gene name.</span></li><li class="c3 c7"><span>Value Filter: filters down the plot-able features by value</span></li><li class="c3 c7"><span>Select Feature: provides the filtered down list of plot-able features to select from based on selected filters.</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-1"><li class="c3 c6"><span>Protein</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-2 start"><li class="c3 c7"><span>Gene Filter: filters down the plot-able features by a specific gene. This is an autocomplete search field for a gene name.</span></li><li class="c3 c7"><span>Protein Filter: filters down the plot-able features by protein. This is an autocomplete search field for a protein name.</span></li><li class="c3 c7"><span>Select Feature: provides the filtered down list of plot-able features to select from based on selected filters.</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-1"><li class="c3 c6"><span>Mutation</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-2 start"><li class="c3 c7"><span>Gene Filter: filters down the plot-able features by a specific gene. This is an autocomplete search field for a gene name.</span></li><li class="c3 c7"><span>Value Filter: filters down the plot-able features by mutation value.</span></li><li class="c3 c7"><span>Select Feature: provides the filtered down list of plot-able features to select from based on selected filters.</span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-0 start"><li class="c3 c5"><span>Swap Values: This button allows you to instantly swap the features on the X and Y Axes without having to re-select each feature individually.</span></li><li class="c3 c5"><span>Color By Cohort: This checkbox will override any feature that is in the Color By Feature. It will use the cohorts provided as the legend and Color By Feature.</span></li><li class="c3 c5"><span>Cohorts: This is where you can select one or more cohorts to plot at one time. </span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-1 start"><li class="c3 c6"><span>To add a cohort, select the &ldquo;+ Cohort&rdquo; option underneath the currently selected list of cohorts. This will take you to the cohorts listing panel where you can select a cohort from the list, or use the autocomplete textbox to search in their list of cohorts. </span></li></ul><ul class="c9 lst-kix_aikc1rnd3yfa-0"><li class="c3 c5"><span>When all the settings have been set, you can click &ldquo;Update Plot&rdquo; to regenerate the plot with the new settings.</span></li></ul><h2 class="c0"><a name="h.s7lh6uemro7y"></a><span>3.9 Pairwise Analysis in Plots</span></h2><p class="c3"><span>If the selected features produce meaningful results from a pairwise calculation, then the results will be shown underneath the plot in a table. If not, then there will be text stating that no results were returned from a pairwise calculation.</span></p><p class="c3 c4"><span></span></p>
            <h1 class="c0"><a name="h.7im6bigprni4"></a><span>4. SeqPeek</span><span>&nbsp;</span></h1><p class="c3"><span>SeqPeek is a visualization designed to show mutations on a protein associated to a Gene. It allows users to select multiple cohort and view mutations on a single gene across all of those cohorts.</span></p><h2 class="c0"><a name="h.jbn4g4b2acgb"></a><span>4.1 Creating a SeqPeek Visualization</span></h2><p class="c3"><span>You can create a new SeqPeek visualization from the user dashboard. Select the &ldquo;SeqPeek&rdquo; option from the &ldquo;+ Create&rdquo; menu. This will automatically take you to a SeqPeek page with no settings preselected for you. To make changes, open the Settings panel.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>In the Settings panel, there will be options to select in order to make a new plot.</span></p><ul class="c9 lst-kix_ckan1xem1ba6-0 start"><li class="c3 c5"><span>Gene selection: This is an autocompleting dropdown that will search the database for valid genes the system has data for. </span></li><li class="c3 c5"><span>Cohorts</span><span>: This is where you can select one or more cohorts to plot at one time. </span></li></ul><ul class="c9 lst-kix_ckan1xem1ba6-1 start"><li class="c3 c6"><span>To add a cohort, select the &ldquo;+ Cohort&rdquo; option underneath the currently selected list of cohorts. This will take you to the cohorts listing panel where you can select a cohort from the list, or use the autocomplete textbox to search in their list of cohorts. </span></li></ul><ul class="c9 lst-kix_ckan1xem1ba6-0"><li class="c3 c5"><span>When all the settings have been set, you can click &ldquo;Update Plot&rdquo; to regenerate the plot with the new settings.</span></li></ul><p class="c3 c4"><span></span></p><h2 class="c0"><a name="h.ob8ciz1drzi0"></a><span>4.2 Saving a SeqPeek Visualization</span></h2><p class="c3"><span>To save changes to the visualization, click &ldquo;Save Visualization&rdquo; button. This will prompt you to name your SeqPeek visualization and save. You will be notified if it saved correctly or not.</span></p><p class="c3 c4"><span></span></p><p class="c3 c4"><span></span></p><h2 class="c0"><a name="h.x0tqi4ub2orw"></a><span>4.4 Deleting a SeqPeek Visualization</span></h2><h3 class="c0"><a name="h.ueganlj7q60r"></a><span>4</span><span>.4.1 From User Dashboard</span></h3><p class="c3"><span>Select the SeqPeek visualizations that you wish to delete using the checkboxes next to the visualization. When one or more are selected, the delete button will be active and you can then proceed to deleting them.</span></p><h3 class="c0"><a name="h.op7vrtzh2m0f"></a><span>4.4.2 From SeqPeek Visualization</span></h3><p class="c3"><span>If you are viewing a SeqPeek visualization you created, then you can delete the cohort from the top right menu option. </span></p>
            <h1 class="c0"><a name="h.6rnx5eq5qtaj"></a><span>5. IGV Browser</span></h1><p class="c3"><span>The IGV Browser connects to the Google Genomics API. The ISB-CGC Project has both open and protected datasets. If you are authorized to access protected access datasets, they will appear in the selection options. You must select a dataset, which will display the related readgroupsets. You can select one or more readgroupsets to display. </span></p><h2 class="c0"><a name="h.n0rl7rbi99ek"></a><span>5.5 Accessing the IGV Browser</span></h2><p class="c3"><span>To access the IGV browser, go to the cohort file list page. The file listing table includes a column labelled &ldquo;IGV&rdquo;. For those files that also have a ReadGroupSet ID in Google Genomics, a green check mark will appear in the column. Clicking on that link will take you to the IGV browser with the appropriate dataset and readgroupset preselected.</span></p>
            <h1 class="c0"><a name="h.a1np5snph00b"></a><span>6. Sharing</span></h1><h2 class="c0"><a name="h.acebs0saaiiz"></a><span>6.1 Sharing a cohort</span></h2><p class="c3"><span>A cohort can only be shared by the owner of the cohort. </span></p><p class="c3 c4"><span></span></p><p class="c3"><span>The person the cohort is shared with has read only access. If they would like to be able to edit the cohort, they would first need to make a copy of it. This only copies the list of samples and participants associated to the cohort, and not any additional information that is private to the original owner of the cohort.</span></p><h2 class="c0"><a name="h.ykdi9efjqdg3"></a><span>6.2 Sharing a visualization</span></h2><p class="c3"><span>A visualization can only be shared by the owner of the visualization.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>The person the visualization is shared with has read only access. They may make changes to the plot settings, but they will not be able to save those changes. If they want to be able to save changes, they need to clone the visualization first. Underlying cohorts are shared with the user when the visualization is shared.making changes to those cohorts will require the user to first clone the cohort as noted above.</span></p><p class="c3 c4"><span></span></p><h2 class="c0"><a name="h.xou8sp1faoh3"></a><span>6.3 Sharing a SeqPeek visualization</span></h2><p class="c3"><span>A SeqPeek visualization can only be shared by the owner of the visualization.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>The person the SeqPeek visualization is shared with has read only access. They may make changes to the settings, but they will not be able to save those changes. If they want to be able to save changes, they need to clone the SeqPeek visualization first. Underlying cohorts are shared with the user when the visualization is shared.making changes to those cohorts will require the user to first clone the cohort as noted above.</span></p><p class="c3 c4"><span></span></p>
            <h1 class="c0"><a name="h.6173glqdalig"></a><span>7. General Permissions</span></h1><h2 class="c0"><a name="h.ay45vvav1j1"></a><span>7.1 Owner</span></h2><p class="c3"><span>As owner of a cohort or visualization, you are able to edit and share your cohort.</span></p><h2 class="c0"><a name="h.uf32ue2cqgnx"></a><span>7.2 Reader</span></h2><p class="c3"><span>You are able to view the cohort or visualization. You may be able to make configuration changes to plots in a visualizations, but you will not be able to save those changes.</span></p><p class="c3 c4"><span></span></p><p class="c3"><span>You are able to comment on a cohort or plot and have those comments shared with everyone who can view that cohort or plot. </span></p><p class="c3 c4"><span></span></p><p class="c3"><span>In order to edit and save changes to a cohort or visualization, you will need to make a copy first. This will make you owner of the new copy.</span></p>
        </div>
    </div>
{% endblock %}

{% block js_file %}{% static 'js/dashboard.js' %}{% endblock %}